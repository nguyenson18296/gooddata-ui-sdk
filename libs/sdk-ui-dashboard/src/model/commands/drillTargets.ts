// (C) 2021 GoodData Corporation

import { ObjRef } from "@gooddata/sdk-model";
import { IAvailableDrillTargets } from "@gooddata/sdk-ui";
import { IDashboardCommand } from "./base";

/**
 * Add widget drill targets
 *
 * @alpha
 */
export interface AddDrillTargets extends IDashboardCommand {
    readonly type: "GDC.DASH/CMD.DRILL_TARGETS.ADD";
    readonly payload: {
        readonly widgetRef: ObjRef;
        readonly availableDrillTargets: IAvailableDrillTargets;
    };
}

/**
 * Create AddDrillTargets {@link AddDrillTargets} command.
 *
 * @param widgetRef - Unique widget ref
 * @param availableDrillTargets - Available widget drill targets {@link @gooddata/sdk-ui#IAvailableDrillTargets}
 * @param correlationId - optionally specify correlation id to use for this command. this will be included in all
 * @returns AddDrillTargets command
 *
 * @alpha
 */
export function addDrillTargets(
    widgetRef: ObjRef,
    availableDrillTargets: IAvailableDrillTargets,
    correlationId?: string,
): AddDrillTargets {
    return {
        type: "GDC.DASH/CMD.DRILL_TARGETS.ADD",
        correlationId,
        payload: {
            widgetRef,
            availableDrillTargets,
        },
    };
}
