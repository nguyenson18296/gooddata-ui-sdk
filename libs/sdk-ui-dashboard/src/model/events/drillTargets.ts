// (C) 2021 GoodData Corporation
import { ObjRef } from "@gooddata/sdk-model";
import { IAvailableDrillTargets } from "@gooddata/sdk-ui";
import { DashboardContext } from "../types/commonTypes";
import { IDashboardEvent } from "./base";

/**
 *  Widget drill targets added event
 *
 * @alpha
 */
export interface DrillTargetsAdded extends IDashboardEvent {
    readonly type: "GDC.DASH/EVT.DRILL_TARGETS.ADDED";
    readonly payload: {
        readonly widgetRef: ObjRef;
        readonly availableDrillTargets: IAvailableDrillTargets;
    };
}

/**
 * Create DrillTargetsAdded {@link DrillTargetsAdded} event.
 *
 * @param widgetRef - Unique widget ref
 * @param availableDrillTargets - Available widget drill targets {@link @gooddata/sdk-ui#IAvailableDrillTargets}
 * @param  correlationId - correlationId
 * @returns - DrillTargetsAdded command
 *
 * @alpha
 */
export function drillTargetsAdded(
    ctx: DashboardContext,
    widgetRef: ObjRef,
    availableDrillTargets: IAvailableDrillTargets,
    correlationId?: string,
): DrillTargetsAdded {
    return {
        type: "GDC.DASH/EVT.DRILL_TARGETS.ADDED",
        ctx,
        correlationId,
        payload: {
            widgetRef,
            availableDrillTargets,
        },
    };
}
