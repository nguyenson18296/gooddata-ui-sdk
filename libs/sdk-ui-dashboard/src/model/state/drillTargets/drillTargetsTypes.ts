// (C) 2021 GoodData Corporation
import { ObjRef } from "@gooddata/sdk-model";
import { IAvailableDrillTargets } from "@gooddata/sdk-ui";

/**
 * Represent state item of widget reported available drill targets
 * @internal
 */
export interface IDrillTargets {
    /**
     * widget ref
     */
    ref: ObjRef;

    /**
     * widget reported available drill targets
     */
    availableDrillTargets?: IAvailableDrillTargets;
}
