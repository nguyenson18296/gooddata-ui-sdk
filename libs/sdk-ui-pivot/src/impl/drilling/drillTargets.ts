// (C) 2007-2021 GoodData Corporation
import {
    DataViewFacade,
    IAvailableDrillTargetAttribute,
    IAvailableDrillTargetMeasure,
    IAvailableDrillTargets,
} from "@gooddata/sdk-ui";
import { IMeasureDescriptor } from "@gooddata/sdk-backend-spi";

export function getAvailableDrillTargets(dv: DataViewFacade): IAvailableDrillTargets {
    const measureDescriptors = dv
        .meta()
        .measureDescriptors()
        .map(
            (measure: IMeasureDescriptor): IAvailableDrillTargetMeasure => ({
                measure,
                attributes: dv.meta().attributeDescriptors(),
            }),
        );

    const rowAttributeItems: IAvailableDrillTargetAttribute[] = dv
        .meta()
        .attributeDescriptorsForDim(0)
        .map((attribute, _index, attributes) => ({
            attribute,
            intersectionAttributes: attributes,
        }));

    return {
        measures: measureDescriptors,
        attributes: rowAttributeItems,
    };
}
