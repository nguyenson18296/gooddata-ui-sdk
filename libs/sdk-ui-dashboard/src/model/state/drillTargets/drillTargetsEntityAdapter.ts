// (C) 2021 GoodData Corporation
import { createEntityAdapter } from "@reduxjs/toolkit";
import { serializeObjRef } from "@gooddata/sdk-model";
import { IDrillTargets } from "./drillTargetsTypes";

export const drillTargetsAdapter = createEntityAdapter<IDrillTargets>({
    selectId: (drillTargets) => serializeObjRef(drillTargets.ref),
});
