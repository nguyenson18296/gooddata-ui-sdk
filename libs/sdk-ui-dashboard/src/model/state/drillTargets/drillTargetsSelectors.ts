// (C) 2021 GoodData Corporation

import { DashboardState } from "../types";
import { drillTargetsAdapter } from "./drillTargetsEntityAdapter";
import memoize from "lodash/memoize";
import { createSelector } from "@reduxjs/toolkit";
import { ObjRef, serializeObjRef } from "@gooddata/sdk-model";
import { newMapForObjectWithIdentity } from "../../../_staging/metadata/objRefMap";

const entitySelectors = drillTargetsAdapter.getSelectors((state: DashboardState) => state.drillTargets);

const selectDrillTargetsInternal = entitySelectors.selectAll;

/**
 * Return all widgets drill targets
 * @alpha
 */
export const selectDrillTargets = createSelector(selectDrillTargetsInternal, (drillTargets) => {
    return newMapForObjectWithIdentity(drillTargets);
});

/**
 * Selects drill targets by widget ref.
 *
 * @alpha
 */
export const selectDrillTargetsByWidgetRef = memoize((ref: ObjRef) => {
    return createSelector(selectDrillTargets, (drillTargets) => {
        return drillTargets.get(ref);
    });
}, serializeObjRef);
