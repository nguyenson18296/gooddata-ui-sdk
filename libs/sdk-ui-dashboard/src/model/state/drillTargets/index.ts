// (C) 2021 GoodData Corporation
import { createSlice } from "@reduxjs/toolkit";
import { drillTargetsAdapter } from "./drillTargetsEntityAdapter";

const drillTargetsSlice = createSlice({
    name: "drillTargets",
    initialState: drillTargetsAdapter.getInitialState(),
    reducers: {
        addDrillTargets: drillTargetsAdapter.addOne,
    },
});

export const drillTargetsReducer = drillTargetsSlice.reducer;

export const drillTargetsActions = drillTargetsSlice.actions;
