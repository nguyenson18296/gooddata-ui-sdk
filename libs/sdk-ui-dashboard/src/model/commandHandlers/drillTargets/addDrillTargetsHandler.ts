// (C) 2021 GoodData Corporation
import { SagaIterator } from "redux-saga";
import { put } from "redux-saga/effects";
import { AddDrillTargets } from "../../commands/drillTargets";
import { DrillTargetsAdded, drillTargetsAdded } from "../../events/drillTargets";
import { internalErrorOccurred } from "../../events/general";
import { drillTargetsActions } from "../../state/drillTargets";
import { DashboardContext } from "../../types/commonTypes";

export function* addDrillTargetsHandler(
    ctx: DashboardContext,
    cmd: AddDrillTargets,
): SagaIterator<DrillTargetsAdded> {
    // eslint-disable-next-line no-console
    console.debug("handling add drill targets", cmd, "in context", ctx);

    try {
        const { widgetRef, availableDrillTargets } = cmd.payload;

        yield put(drillTargetsActions.addDrillTargets({ ref: widgetRef, availableDrillTargets }));
        return drillTargetsAdded(ctx, widgetRef, availableDrillTargets, cmd.correlationId);
    } catch (e) {
        throw internalErrorOccurred(
            ctx,
            "An unexpected error has occurred while adding drill targets",
            e,
            cmd.correlationId,
        );
    }
}
