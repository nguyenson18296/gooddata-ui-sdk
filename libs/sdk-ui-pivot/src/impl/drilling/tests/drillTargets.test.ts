// (C) 2021 GoodData Corporation
import {
    TwoMeasuresWithTwoRowAndTwoColumnAttributes,
    OneMeasureAndRepeatingRowAttributesOnDifferentPositions,
} from "./drilling.fixture";
import { getAvailableDrillTargets } from "../drillTargets";

describe("getAvailableDrillTargets", () => {
    it("gets propper available drill targets - ignoring column attributes", () => {
        expect(getAvailableDrillTargets(TwoMeasuresWithTwoRowAndTwoColumnAttributes)).toMatchSnapshot();
    });

    it("gets propper available drill targets for duplicated row attribute", () => {
        expect(
            getAvailableDrillTargets(OneMeasureAndRepeatingRowAttributesOnDifferentPositions),
        ).toMatchSnapshot();
    });
});
