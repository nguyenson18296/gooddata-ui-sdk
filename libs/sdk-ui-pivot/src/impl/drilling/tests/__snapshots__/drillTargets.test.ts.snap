// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getAvailableDrillTargets gets propper available drill targets - ignoring column attributes 1`] = `
Object {
  "attributes": Array [
    Object {
      "attribute": Object {
        "attributeHeader": Object {
          "formOf": Object {
            "identifier": "attr.product.id",
            "name": "Product",
            "ref": Object {
              "uri": "/gdc/md/referenceworkspace/obj/1054",
            },
            "uri": "/gdc/md/referenceworkspace/obj/1054",
          },
          "identifier": "label.product.id.name",
          "localIdentifier": "a_label.product.id.name",
          "name": "Product Name",
          "ref": Object {
            "uri": "/gdc/md/referenceworkspace/obj/1055",
          },
          "uri": "/gdc/md/referenceworkspace/obj/1055",
        },
      },
      "intersectionAttributes": Array [
        Object {
          "attributeHeader": Object {
            "formOf": Object {
              "identifier": "attr.product.id",
              "name": "Product",
              "ref": Object {
                "uri": "/gdc/md/referenceworkspace/obj/1054",
              },
              "uri": "/gdc/md/referenceworkspace/obj/1054",
            },
            "identifier": "label.product.id.name",
            "localIdentifier": "a_label.product.id.name",
            "name": "Product Name",
            "ref": Object {
              "uri": "/gdc/md/referenceworkspace/obj/1055",
            },
            "uri": "/gdc/md/referenceworkspace/obj/1055",
          },
        },
      ],
    },
    Object {
      "attribute": Object {
        "attributeHeader": Object {
          "formOf": Object {
            "identifier": "attr.owner.department",
            "name": "Department",
            "ref": Object {
              "uri": "/gdc/md/referenceworkspace/obj/1088",
            },
            "uri": "/gdc/md/referenceworkspace/obj/1088",
          },
          "identifier": "label.owner.department",
          "localIdentifier": "a_label.owner.department",
          "name": "Department",
          "ref": Object {
            "uri": "/gdc/md/referenceworkspace/obj/1089",
          },
          "uri": "/gdc/md/referenceworkspace/obj/1089",
        },
      },
      "intersectionAttributes": Array [
        Object {
          "attributeHeader": Object {
            "formOf": Object {
              "identifier": "attr.product.id",
              "name": "Product",
              "ref": Object {
                "uri": "/gdc/md/referenceworkspace/obj/1054",
              },
              "uri": "/gdc/md/referenceworkspace/obj/1054",
            },
            "identifier": "label.product.id.name",
            "localIdentifier": "a_label.product.id.name",
            "name": "Product Name",
            "ref": Object {
              "uri": "/gdc/md/referenceworkspace/obj/1055",
            },
            "uri": "/gdc/md/referenceworkspace/obj/1055",
          },
        },
        Object {
          "attributeHeader": Object {
            "formOf": Object {
              "identifier": "attr.owner.department",
              "name": "Department",
              "ref": Object {
                "uri": "/gdc/md/referenceworkspace/obj/1088",
              },
              "uri": "/gdc/md/referenceworkspace/obj/1088",
            },
            "identifier": "label.owner.department",
            "localIdentifier": "a_label.owner.department",
            "name": "Department",
            "ref": Object {
              "uri": "/gdc/md/referenceworkspace/obj/1089",
            },
            "uri": "/gdc/md/referenceworkspace/obj/1089",
          },
        },
      ],
    },
  ],
  "measures": Array [
    Object {
      "attributes": Array [
        Object {
          "attributeHeader": Object {
            "formOf": Object {
              "identifier": "attr.product.id",
              "name": "Product",
              "ref": Object {
                "uri": "/gdc/md/referenceworkspace/obj/1054",
              },
              "uri": "/gdc/md/referenceworkspace/obj/1054",
            },
            "identifier": "label.product.id.name",
            "localIdentifier": "a_label.product.id.name",
            "name": "Product Name",
            "ref": Object {
              "uri": "/gdc/md/referenceworkspace/obj/1055",
            },
            "uri": "/gdc/md/referenceworkspace/obj/1055",
          },
        },
        Object {
          "attributeHeader": Object {
            "formOf": Object {
              "identifier": "attr.owner.department",
              "name": "Department",
              "ref": Object {
                "uri": "/gdc/md/referenceworkspace/obj/1088",
              },
              "uri": "/gdc/md/referenceworkspace/obj/1088",
            },
            "identifier": "label.owner.department",
            "localIdentifier": "a_label.owner.department",
            "name": "Department",
            "ref": Object {
              "uri": "/gdc/md/referenceworkspace/obj/1089",
            },
            "uri": "/gdc/md/referenceworkspace/obj/1089",
          },
        },
        Object {
          "attributeHeader": Object {
            "formOf": Object {
              "identifier": "attr.opportunitysnapshot.forecastcategory",
              "name": "Forecast Category",
              "ref": Object {
                "uri": "/gdc/md/referenceworkspace/obj/1062",
              },
              "uri": "/gdc/md/referenceworkspace/obj/1062",
            },
            "identifier": "label.opportunitysnapshot.forecastcategory",
            "localIdentifier": "a_label.opportunitysnapshot.forecastcategory",
            "name": "Forecast Category",
            "ref": Object {
              "uri": "/gdc/md/referenceworkspace/obj/1063",
            },
            "uri": "/gdc/md/referenceworkspace/obj/1063",
          },
        },
        Object {
          "attributeHeader": Object {
            "formOf": Object {
              "identifier": "attr.owner.region",
              "name": "Region",
              "ref": Object {
                "uri": "/gdc/md/referenceworkspace/obj/1086",
              },
              "uri": "/gdc/md/referenceworkspace/obj/1086",
            },
            "identifier": "label.owner.region",
            "localIdentifier": "a_label.owner.region",
            "name": "Region",
            "ref": Object {
              "uri": "/gdc/md/referenceworkspace/obj/1087",
            },
            "uri": "/gdc/md/referenceworkspace/obj/1087",
          },
        },
      ],
      "measure": Object {
        "measureHeaderItem": Object {
          "format": "$#,##0.00",
          "identifier": "aangOxLSeztu",
          "localIdentifier": "m_aangOxLSeztu",
          "name": "Amount",
          "ref": Object {
            "uri": "/gdc/md/referenceworkspace/obj/1271",
          },
          "uri": "/gdc/md/referenceworkspace/obj/1271",
        },
      },
    },
    Object {
      "attributes": Array [
        Object {
          "attributeHeader": Object {
            "formOf": Object {
              "identifier": "attr.product.id",
              "name": "Product",
              "ref": Object {
                "uri": "/gdc/md/referenceworkspace/obj/1054",
              },
              "uri": "/gdc/md/referenceworkspace/obj/1054",
            },
            "identifier": "label.product.id.name",
            "localIdentifier": "a_label.product.id.name",
            "name": "Product Name",
            "ref": Object {
              "uri": "/gdc/md/referenceworkspace/obj/1055",
            },
            "uri": "/gdc/md/referenceworkspace/obj/1055",
          },
        },
        Object {
          "attributeHeader": Object {
            "formOf": Object {
              "identifier": "attr.owner.department",
              "name": "Department",
              "ref": Object {
                "uri": "/gdc/md/referenceworkspace/obj/1088",
              },
              "uri": "/gdc/md/referenceworkspace/obj/1088",
            },
            "identifier": "label.owner.department",
            "localIdentifier": "a_label.owner.department",
            "name": "Department",
            "ref": Object {
              "uri": "/gdc/md/referenceworkspace/obj/1089",
            },
            "uri": "/gdc/md/referenceworkspace/obj/1089",
          },
        },
        Object {
          "attributeHeader": Object {
            "formOf": Object {
              "identifier": "attr.opportunitysnapshot.forecastcategory",
              "name": "Forecast Category",
              "ref": Object {
                "uri": "/gdc/md/referenceworkspace/obj/1062",
              },
              "uri": "/gdc/md/referenceworkspace/obj/1062",
            },
            "identifier": "label.opportunitysnapshot.forecastcategory",
            "localIdentifier": "a_label.opportunitysnapshot.forecastcategory",
            "name": "Forecast Category",
            "ref": Object {
              "uri": "/gdc/md/referenceworkspace/obj/1063",
            },
            "uri": "/gdc/md/referenceworkspace/obj/1063",
          },
        },
        Object {
          "attributeHeader": Object {
            "formOf": Object {
              "identifier": "attr.owner.region",
              "name": "Region",
              "ref": Object {
                "uri": "/gdc/md/referenceworkspace/obj/1086",
              },
              "uri": "/gdc/md/referenceworkspace/obj/1086",
            },
            "identifier": "label.owner.region",
            "localIdentifier": "a_label.owner.region",
            "name": "Region",
            "ref": Object {
              "uri": "/gdc/md/referenceworkspace/obj/1087",
            },
            "uri": "/gdc/md/referenceworkspace/obj/1087",
          },
        },
      ],
      "measure": Object {
        "measureHeaderItem": Object {
          "format": "$#,##0.00",
          "identifier": "acugFHNJgsBy",
          "localIdentifier": "m_acugFHNJgsBy",
          "name": "Won",
          "ref": Object {
            "uri": "/gdc/md/referenceworkspace/obj/1272",
          },
          "uri": "/gdc/md/referenceworkspace/obj/1272",
        },
      },
    },
  ],
}
`;

exports[`getAvailableDrillTargets gets propper available drill targets for duplicated row attribute 1`] = `
Object {
  "attributes": Array [
    Object {
      "attribute": Object {
        "attributeHeader": Object {
          "formOf": Object {
            "identifier": "attr.product.id",
            "name": "Product",
            "ref": Object {
              "uri": "/gdc/md/referenceworkspace/obj/1054",
            },
            "uri": "/gdc/md/referenceworkspace/obj/1054",
          },
          "identifier": "label.product.id.name",
          "localIdentifier": "a_label.product.id.name",
          "name": "Product Name",
          "ref": Object {
            "uri": "/gdc/md/referenceworkspace/obj/1055",
          },
          "uri": "/gdc/md/referenceworkspace/obj/1055",
        },
      },
      "intersectionAttributes": Array [
        Object {
          "attributeHeader": Object {
            "formOf": Object {
              "identifier": "attr.product.id",
              "name": "Product",
              "ref": Object {
                "uri": "/gdc/md/referenceworkspace/obj/1054",
              },
              "uri": "/gdc/md/referenceworkspace/obj/1054",
            },
            "identifier": "label.product.id.name",
            "localIdentifier": "a_label.product.id.name",
            "name": "Product Name",
            "ref": Object {
              "uri": "/gdc/md/referenceworkspace/obj/1055",
            },
            "uri": "/gdc/md/referenceworkspace/obj/1055",
          },
        },
      ],
    },
    Object {
      "attribute": Object {
        "attributeHeader": Object {
          "formOf": Object {
            "identifier": "attr.owner.region",
            "name": "Region",
            "ref": Object {
              "uri": "/gdc/md/referenceworkspace/obj/1086",
            },
            "uri": "/gdc/md/referenceworkspace/obj/1086",
          },
          "identifier": "label.owner.region",
          "localIdentifier": "a_label.owner.region",
          "name": "Region",
          "ref": Object {
            "uri": "/gdc/md/referenceworkspace/obj/1087",
          },
          "uri": "/gdc/md/referenceworkspace/obj/1087",
        },
      },
      "intersectionAttributes": Array [
        Object {
          "attributeHeader": Object {
            "formOf": Object {
              "identifier": "attr.product.id",
              "name": "Product",
              "ref": Object {
                "uri": "/gdc/md/referenceworkspace/obj/1054",
              },
              "uri": "/gdc/md/referenceworkspace/obj/1054",
            },
            "identifier": "label.product.id.name",
            "localIdentifier": "a_label.product.id.name",
            "name": "Product Name",
            "ref": Object {
              "uri": "/gdc/md/referenceworkspace/obj/1055",
            },
            "uri": "/gdc/md/referenceworkspace/obj/1055",
          },
        },
        Object {
          "attributeHeader": Object {
            "formOf": Object {
              "identifier": "attr.owner.region",
              "name": "Region",
              "ref": Object {
                "uri": "/gdc/md/referenceworkspace/obj/1086",
              },
              "uri": "/gdc/md/referenceworkspace/obj/1086",
            },
            "identifier": "label.owner.region",
            "localIdentifier": "a_label.owner.region",
            "name": "Region",
            "ref": Object {
              "uri": "/gdc/md/referenceworkspace/obj/1087",
            },
            "uri": "/gdc/md/referenceworkspace/obj/1087",
          },
        },
      ],
    },
    Object {
      "attribute": Object {
        "attributeHeader": Object {
          "formOf": Object {
            "identifier": "attr.product.id",
            "name": "Product",
            "ref": Object {
              "uri": "/gdc/md/referenceworkspace/obj/1054",
            },
            "uri": "/gdc/md/referenceworkspace/obj/1054",
          },
          "identifier": "label.product.id.name",
          "localIdentifier": "product.name.test",
          "name": "Product Name",
          "ref": Object {
            "uri": "/gdc/md/referenceworkspace/obj/1055",
          },
          "uri": "/gdc/md/referenceworkspace/obj/1055",
        },
      },
      "intersectionAttributes": Array [
        Object {
          "attributeHeader": Object {
            "formOf": Object {
              "identifier": "attr.product.id",
              "name": "Product",
              "ref": Object {
                "uri": "/gdc/md/referenceworkspace/obj/1054",
              },
              "uri": "/gdc/md/referenceworkspace/obj/1054",
            },
            "identifier": "label.product.id.name",
            "localIdentifier": "a_label.product.id.name",
            "name": "Product Name",
            "ref": Object {
              "uri": "/gdc/md/referenceworkspace/obj/1055",
            },
            "uri": "/gdc/md/referenceworkspace/obj/1055",
          },
        },
        Object {
          "attributeHeader": Object {
            "formOf": Object {
              "identifier": "attr.owner.region",
              "name": "Region",
              "ref": Object {
                "uri": "/gdc/md/referenceworkspace/obj/1086",
              },
              "uri": "/gdc/md/referenceworkspace/obj/1086",
            },
            "identifier": "label.owner.region",
            "localIdentifier": "a_label.owner.region",
            "name": "Region",
            "ref": Object {
              "uri": "/gdc/md/referenceworkspace/obj/1087",
            },
            "uri": "/gdc/md/referenceworkspace/obj/1087",
          },
        },
        Object {
          "attributeHeader": Object {
            "formOf": Object {
              "identifier": "attr.product.id",
              "name": "Product",
              "ref": Object {
                "uri": "/gdc/md/referenceworkspace/obj/1054",
              },
              "uri": "/gdc/md/referenceworkspace/obj/1054",
            },
            "identifier": "label.product.id.name",
            "localIdentifier": "product.name.test",
            "name": "Product Name",
            "ref": Object {
              "uri": "/gdc/md/referenceworkspace/obj/1055",
            },
            "uri": "/gdc/md/referenceworkspace/obj/1055",
          },
        },
      ],
    },
  ],
  "measures": Array [
    Object {
      "attributes": Array [
        Object {
          "attributeHeader": Object {
            "formOf": Object {
              "identifier": "attr.product.id",
              "name": "Product",
              "ref": Object {
                "uri": "/gdc/md/referenceworkspace/obj/1054",
              },
              "uri": "/gdc/md/referenceworkspace/obj/1054",
            },
            "identifier": "label.product.id.name",
            "localIdentifier": "a_label.product.id.name",
            "name": "Product Name",
            "ref": Object {
              "uri": "/gdc/md/referenceworkspace/obj/1055",
            },
            "uri": "/gdc/md/referenceworkspace/obj/1055",
          },
        },
        Object {
          "attributeHeader": Object {
            "formOf": Object {
              "identifier": "attr.owner.region",
              "name": "Region",
              "ref": Object {
                "uri": "/gdc/md/referenceworkspace/obj/1086",
              },
              "uri": "/gdc/md/referenceworkspace/obj/1086",
            },
            "identifier": "label.owner.region",
            "localIdentifier": "a_label.owner.region",
            "name": "Region",
            "ref": Object {
              "uri": "/gdc/md/referenceworkspace/obj/1087",
            },
            "uri": "/gdc/md/referenceworkspace/obj/1087",
          },
        },
        Object {
          "attributeHeader": Object {
            "formOf": Object {
              "identifier": "attr.product.id",
              "name": "Product",
              "ref": Object {
                "uri": "/gdc/md/referenceworkspace/obj/1054",
              },
              "uri": "/gdc/md/referenceworkspace/obj/1054",
            },
            "identifier": "label.product.id.name",
            "localIdentifier": "product.name.test",
            "name": "Product Name",
            "ref": Object {
              "uri": "/gdc/md/referenceworkspace/obj/1055",
            },
            "uri": "/gdc/md/referenceworkspace/obj/1055",
          },
        },
      ],
      "measure": Object {
        "measureHeaderItem": Object {
          "format": "$#,##0.00",
          "identifier": "aangOxLSeztu",
          "localIdentifier": "m_aangOxLSeztu",
          "name": "Amount",
          "ref": Object {
            "uri": "/gdc/md/referenceworkspace/obj/1271",
          },
          "uri": "/gdc/md/referenceworkspace/obj/1271",
        },
      },
    },
  ],
}
`;
